### PR TITLE
feat(flask): trace streamed responses

### PR DESCRIPTION
**Pull request by mabdinur**
## Description

In an instrumented flask application ddtrace does not include spans generated by a streamed response in the root `flask.request` span. Currently the `flask.request` span terminates before the first byte is streamed to the client. 

To address this issue this proposal wraps flask applications in a wsgi middleware. The `flask.request` span produced by this middleware will encompass both receiving a request and sending the response to a client.

Screen Shot 2022-06-28 at 4 46 08 PM

## Checklist
- [x] Title must conform to [conventional commit](https://github.com/conventional-changelog/commitlint/tree/master/%40commitlint/config-conventional).
- [x] Add additional sections for `feat` and `fix` pull requests.
- [x] Ensure tests are passing for affected code.
- [x] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Motivation

Accurately measure the duration of flask request. Instrument streamed/buffered responses. 

## Design 

Use the _BaseWSGIMiddleware introduced in this change: https://github.com/DataDog/dd-trace-py/pull/3907. By extending this base class we will wrap the flask request, the flask wsgi_app, and the response in a span. 

Note - This is not a breaking change. The top level flask span `flask.request` will have the same tags and attributes (resource name, service name, error fields, etc). This PR introduces adds more visibility, it does not remove an existing functionality

## Testing strategy

1. create a flask app with an endpoint that returns a generator.
2. run the flask app and send a request to the endpoint
3. flush all spans to the agent
4. ensure the following spans are generated:
  - flask.request
  - flask.application 
  - flask.response

## Reviewer Checklist
- [ ] Title is accurate.
- [ ] Description motivates each change.
- [ ] No unnecessary changes were introduced in this PR.
- [ ] PR cannot be broken up into smaller PRs.
- [ ] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [ ] All relevant GitHub issues are correctly linked.
- [ ] Backports are identified and tagged with Mergifyio.
- [ ] Add to milestone.

### CODE REVIEW

File: ddtrace/contrib/flask/patch.py
Comment (mabdinur) [inline ddtrace/contrib/flask/patch.py:303]: I don't think this span ever closes 😞

File: ddtrace/contrib/flask/patch.py
Comment (brettlangdon) [inline ddtrace/contrib/flask/patch.py:316]:
```suggestion
        with pin.tracer.trace(
            "flask.streamed_response",
            service=trace_utils.int_service(pin, config.flask),
            resource=resource,
            span_type=SpanTypes.WEB,
        ):
            while True:
                yield iterable.__next__()
```


Does something like this work?

Or well, we probably need to ignore `StopIteration` if we do this?

or maybe:

```python
span = pin.tracer.trace(...)
try:
    while True:
        yield iterable.__next__()
finally:
    span.finish()
```

Which would ignore any exceptions raised, but is that what we'd want? 🤔

File: ddtrace/contrib/flask/patch.py
Comment (brettlangdon) [inline ddtrace/contrib/flask/patch.py:126]: do we need both?

File: ddtrace/contrib/flask/patch.py
Comment (brettlangdon) [inline ddtrace/contrib/flask/patch.py:299]: we probably don't _need_ a resource name here. wdyt?

File: ddtrace/contrib/flask/patch.py
Comment (brettlangdon) [inline ddtrace/contrib/flask/patch.py:302]: We should be getting a `StopIteration` exception raised from `iterable.__next__()` when there is no more to iterate over, right?

File: ddtrace/contrib/flask/patch.py
Comment (mabdinur) [reply ddtrace/contrib/flask/patch.py]: Both would work. I just went with the first approach to catch Exceptions

File: ddtrace/contrib/flask/patch.py
Comment (brettlangdon) [reply ddtrace/contrib/flask/patch.py:302]: with the `with span: while True`, do we end up with always marking this span as an error?

File: ddtrace/contrib/flask/patch.py
Comment (mabdinur) [reply ddtrace/contrib/flask/patch.py:126]: Yes because `stream_with_context` can be imported from both modules. Although it seems like there's only documentation for [flask.stream_with_context](https://flask.palletsprojects.com/en/2.1.x/patterns/streaming/#streaming-with-context) so we could just support that function import. I have no strong opinions here

File: ddtrace/contrib/flask/patch.py
Comment (brettlangdon) [reply ddtrace/contrib/flask/patch.py:319]: oh, no... it isn't.... but.... still not sure if that is what we want?

if this try/except was inside the `while True`, then it would cause an infinite loop, but then the way this is our `Stopiteration` will get raised instead of `iterable`s `StopIteration` which might be fine... ?

File: tests/snapshots/tests.contrib.flask.test_request.test_request_streaming_with_context.json
Comment (mabdinur) [inline tests/snapshots/tests.contrib.flask.test_request.test_request_streaming_with_context.json:103]: Note - traced_function_1 was not generated. Should we still use the test client?

File: ddtrace/contrib/flask/patch.py
Comment (mabdinur) [inline ddtrace/contrib/flask/patch.py:286]: Avoid using stream_with_context. Provide support for different wsgi streamed response.

File: ddtrace/contrib/flask/patch.py
Comment (brettlangdon) [inline ddtrace/contrib/flask/patch.py:266]:
```suggestion
    return middleware(environ, start_response)
```

File: tests/contrib/flask/test_request.py
Comment (mabdinur) [reply tests/contrib/flask/test_request.py]: We don't produce any snapshots for any flask tests. I'll stick to the existing convention of using TracerTestCase and the DummyTracer. I tried updating all the flask tests to snapshots but it's a much large effort than I thought: https://github.com/DataDog/dd-trace-py/compare/munir/use-snapshot-tests-in-flask?expand=1.. Currently we're leaking tracer configurations between tests and this causes tests to fail.

**Comment by mabdinur** [comment]
Blocked by: https://github.com/DataDog/dd-trace-py/pull/3907

**Comment by mabdinur** [comment]
Blocked by: https://github.com/DataDog/dd-trace-py/pull/3958. Tracing streamed responses should be opt to maintain backwards compatibility. After this change is merge the _FlaskWSGIMiddleware will override `_instrument_response()` and check the environment variable ``DD_TRACE_FLASK_RESPONSE_ENABLED``

**Comment by mabdinur** [comment]
We can clean up the flask middleware code in a follow up change. Here's what I had in mind: 520d5beb9beb7390028e82b5eb7e1120fc10f44f. I would like to minimize the changes in this PR

File: releasenotes/notes/flask-support-response-streamig-d3aafc8fa1e3605d.yaml
Comment (brettlangdon) [inline releasenotes/notes/flask-support-response-streamig-d3aafc8fa1e3605d.yaml:4]:
```suggestion
    flask: Update the duration of a flask request to include streamed response. Previously ``flask.request`` spans would finish before the first byte was sent to the client.
```

**Review by brettlangdon — APPROVED** [review]

**Review by brettlangdon — APPROVED** [review]

File: ddtrace/contrib/flask/patch.py
Comment (Kyle-Verhoog) [inline ddtrace/contrib/flask/patch.py:110]: Stale, this function is removed now

File: ddtrace/contrib/flask/patch.py
Comment (Kyle-Verhoog) [inline ddtrace/contrib/flask/patch.py:111]: `set_http_meta` will strip the query string if present, but this was true before this PR too, so no need to act on this now.

File: ddtrace/contrib/flask/patch.py
Comment (Kyle-Verhoog) [inline ddtrace/contrib/flask/patch.py:67]: This comment seems pretty stale here now. It made more sense when the function was being wrapped.

File: releasenotes/notes/flask-support-response-streamig-d3aafc8fa1e3605d.yaml
Comment (Kyle-Verhoog) [inline releasenotes/notes/flask-support-response-streamig-d3aafc8fa1e3605d.yaml:4]:
```suggestion
    flask: update the duration of a flask request to include streamed responses. Previously ``flask.request`` spans would finish before the first byte was sent to the client.
```

**Review by Kyle-Verhoog — CHANGES_REQUESTED** [review]

File: ddtrace/contrib/flask/patch.py
Comment (brettlangdon) [inline ddtrace/contrib/flask/patch.py:111]:
```suggestion
```

File: ddtrace/contrib/flask/patch.py
Comment (brettlangdon) [reply ddtrace/contrib/flask/patch.py:67]:
```suggestion
```

File: ddtrace/contrib/flask/patch.py
Comment (Kyle-Verhoog) [inline ddtrace/contrib/flask/patch.py:68]: This comment is also stale, we no longer patch flask's start_response, the comment still holds up otherwise though

File: ddtrace/contrib/flask/patch.py
Comment (Kyle-Verhoog) [inline ddtrace/contrib/flask/patch.py:70]:
```suggestion
```

File: tests/contrib/flask/test_errorhandler.py
Comment (Kyle-Verhoog) [inline tests/contrib/flask/test_errorhandler.py:15]: would probably be good to introduce a snapshot test case before this PR to see the real impact on the traces produced

Comment (Kyle-Verhoog) [reply tests/contrib/flask/test_errorhandler.py:15]: Done in #3971

**Review by Kyle-Verhoog — DISMISSED** [review]

**Review by brettlangdon — APPROVED** [review]

**Comment by mabdinur** [comment]
failing flask snapshots: https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/20152/workflows/0e7c9f4e-80cb-4f3a-ab5c-1ea63afa7d8f/jobs/1364413

**Review by Kyle-Verhoog — DISMISSED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by brettlangdon — APPROVED** [review]

**Pull request merged**
